### PR TITLE
James 1712 2

### PR DESCRIPTION
**Pull request by aduprat**
Rebase #198 and take care of comments.
Two new tests added: 
- creating a mailbox that already exists
- cycle detection

### CODE REVIEW

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java [hunk 1]
```diff
@@ -19,20 +19,120 @@
...
+            Map<MailboxCreationId, String> creationIdsToCreatedMailboxId, SetMailboxesResponse.Builder builder) {
+        try {
+            MailboxPath mailboxPath = getMailboxPath(mailboxRequest, creationIdsToCreatedMailboxId, mailboxSession);
+            mailboxManager.createMailbox(mailboxPath, mailboxSession);
+            Optional<Mailbox> mailbox = mailboxUtils.mailboxFromMailboxPath(mailboxPath, mailboxSession);
+            if (mailbox.isPresent()) {
+                builder.creation(mailboxCreationId, mailbox.get());
+                creationIdsToCreatedMailboxId.put(mailboxCreationId, mailbox.get().getId());
+            } else {
+                builder.notCreated(mailboxCreationId, SetError.builder()
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java]: When does this occur?

Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java]: When the created mailbox has been deleted between `createMailbox` and `mailboxFromMailboxPath` calls.

Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java]: Why do you need to check the creation? A non throwing mailboxManager.createMailbox is not sufficient?

Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java]: As the `MailboxManager.createMailbox` returns nothing, i have to get the mailbox after creation in order to send it to the response.
The `MailboxManager.createMailbox` method is not sexy; i don't really want to make it return something:
```java
    public void createMailbox(MailboxPath mailboxPath, final MailboxSession mailboxSession)
            throws MailboxException {
        mailboxSession.getLog().debug("createMailbox " + mailboxPath);
        final int length = mailboxPath.getName().length();
        if (length == 0) {
            mailboxSession.getLog().warn("Ignoring mailbox with empty name");
        } else {
            if (mailboxPath.getName().charAt(length - 1) == getDelimiter())
                mailboxPath.setName(mailboxPath.getName().substring(0, length - 1));
            if (mailboxExists(mailboxPath, mailboxSession))
                throw new MailboxExistsException(mailboxPath.toString());
            // Create parents first
            // If any creation fails then the mailbox will not be created
            // TODO: transaction
            for (final MailboxPath mailbox : mailboxPath.getHierarchyLevels(getDelimiter()))

                locker.executeWithLock(mailboxSession, mailbox, new LockAwareExecution<Void>() {

                    public Void execute() throws MailboxException {
                        if (!mailboxExists(mailbox, mailboxSession)) {
                            final org.apache.james.mailbox.store.mail.model.Mailbox<Id> m = doCreateMailbox(mailbox, mailboxSession);
                            final MailboxMapper<Id> mapper = mailboxSessionMapperFactory.getMailboxMapper(mailboxSession);
                            mapper.execute(new TransactionalMapper.VoidTransaction() {

                                public void runVoid() throws MailboxException {
                                    mapper.save(m);
                                }

                            });

                            // notify listeners
                            dispatcher.mailboxAdded(mailboxSession, m);
                        }
                        return null;

                    }
                }, true);

        }
    }
```

Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java]: ok I forgot that create must return the new mailbox.

**Comment by rouazana** [comment]
What's happening when a user tries to create a mailbox which name contains a mailboxpath separator?

**Comment by aduprat** [comment]
We should limit such case.

**Comment by rouazana** [comment]
:+1:

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessorTest.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import org.apache.james.mailbox.MailboxManager;
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.store.TestId;
+import org.junit.Before;
+import org.junit.Test;
+
+public class SetMailboxesCreationProcessorTest {
+
+    private MailboxManager mailboxManager;
```
Comment (fvn-linagora) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessorTest.java]: only used once in Before, can be inlined (ie remove field).

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessorTest.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    private MailboxUtils<TestId> mailboxUtils;
+    private SetMailboxesCreationProcessor<TestId> sut;
+
+    @Before
+    @SuppressWarnings("unchecked")
+    public void setup() {
+        mailboxManager = mock(MailboxManager.class);
+        mailboxUtils = mock(MailboxUtils.class);
+        sut = new SetMailboxesCreationProcessor<TestId>(mailboxManager, mailboxUtils);
```
Comment (fvn-linagora) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessorTest.java]: <TestId> superfluous (<>), inferrred by sut field type declaration.

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessorTest.java (hunk 2 above)
Comment (aduprat) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessorTest.java]: OK

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessorTest.java (hunk 3 above)
Comment (aduprat) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessorTest.java]: OK

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java [hunk 4]
```diff
@@ -19,20 +19,130 @@
...
-                entry.getKey(),
-                Mailbox.builder().name(entry.getValue().getName()).id("serverId").build()));
+        try {
+            Map<MailboxCreationId, String> creationIdsToCreatedMailboxId = new HashMap<>();
+            collectionHierarchySorter.sortFromRootToLeaf(request.getCreate().entrySet())
+                .forEach(entry -> 
+                    createMailbox(entry.getKey(), entry.getValue(), mailboxSession, creationIdsToCreatedMailboxId, builder));
+        } catch (CycleDetectedException e) {
+            request.getCreate().entrySet()
+                .forEach(entry ->
```
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java]: Do we really intend to return as many SetErrors as the number of creation ids in the submitted SetMailboxes request ?
Do we need to iterate when a cycle is found ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java [hunk 5]
```diff
@@ -19,20 +19,130 @@
...
-        request.getCreate().entrySet().stream()
-            .forEach(entry -> builder.creation(
-                entry.getKey(),
-                Mailbox.builder().name(entry.getValue().getName()).id("serverId").build()));
+        try {
+            Map<MailboxCreationId, String> creationIdsToCreatedMailboxId = new HashMap<>();
+            collectionHierarchySorter.sortFromRootToLeaf(request.getCreate().entrySet())
+                .forEach(entry -> 
+                    createMailbox(entry.getKey(), entry.getValue(), mailboxSession, creationIdsToCreatedMailboxId, builder));
+        } catch (CycleDetectedException e) {
```
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java]: can be put the catching closer to the thrower (sortFromRootToLead) to make error handling more readable ?
The code is misleading and let readers think we may create some mailboxes before some cycles could be found (which is thankfully not true!).

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java [hunk 6]
```diff
@@ -19,20 +19,130 @@
...
+            String message = String.format("An error occurred when creating the mailbox '%s'", mailboxCreationId.getCreationId());
+            LOGGER.error(message, e);
+            builder.notCreated(mailboxCreationId, SetError.builder()
+                    .type("anErrorOccurred")
+                    .description(message)
+                    .build());
+        }
+    }
+
+    private void checkName(MailboxRequest mailboxRequest, MailboxSession mailboxSession) {
```
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java]: "ensure" of "assert" are usually more explicit when implementing some sort of preconditions, I'd suggest.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java (hunk 4 above)
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java]: Yes, we don't have to be smart when a cycle is detected -> we will reject all the creation requests.
In JMAP, we have to respond to each requests, so it's normal to iterate on the creation requests.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java (hunk 5 above)
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java]: I don't see the advantage, as i will have to extract the `sortFromRootToLeaf` in an extra variable.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java (hunk 6 above)
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java]: OK

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java (hunk 4 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java]: OK, so maybe extract handling logic into a method ?

Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java]: Ok

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java [hunk 7]
```diff
@@ -19,20 +19,130 @@
...
+        String name = mailboxRequest.getName();
+        char pathDelimiter = mailboxSession.getPathDelimiter();
+        if (name.contains(String.valueOf(pathDelimiter))) {
+            throw new MailboxNameException(String.format("The mailbox '%s' contains an illegal character: '%c'", name, pathDelimiter));
+        }
+    }
+
+    private MailboxPath getMailboxPath(MailboxRequest mailboxRequest, Map<MailboxCreationId, String> creationIdsToCreatedMailboxId, MailboxSession mailboxSession) throws MailboxException {
+        if (mailboxRequest.getParentId().isPresent()) {
```
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java]: using both isPresent() and .get() looks like a smell. I get that throwing calls makes writing an expression more painful than it should.
But can't we get the parentName in a nicer way ?

Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java]: There's no nicer way to get the parent name, IMHO.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/SetMailboxesResponse.java [hunk 8]
```diff
@@ -31,10 +33,12 @@ public static Builder builder() {
...
-        private ImmutableMap.Builder<MailboxCreationId, Mailbox> created;
+        private final ImmutableMap.Builder<MailboxCreationId, Mailbox> created;
+        private final ImmutableMap.Builder<MailboxCreationId, SetError> notCreated;
 
         private Builder() {
             created = ImmutableMap.builder();
+            notCreated = ImmutableMap.builder();
         }
 
         public Builder creation(MailboxCreationId creationId, Mailbox mailbox) {
```
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/SetMailboxesResponse.java]: Why "creation" (along with notCreated) ? Did you mean created ?

Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/SetMailboxesResponse.java]: This is fixed in this commit:
https://github.com/linagora/james-project/pull/202/commits/e1cae25134d1bbb897e6c2d7186ddca392e01796

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/CollectionHierarchySorter.java [hunk 9]
```diff
@@ -19,36 +19,46 @@
...
 import java.util.function.Function;
 import java.util.stream.Collectors;
 
-import org.apache.james.jmap.model.mailbox.Mailbox;
 import org.apache.james.jmap.utils.DependencyGraph.CycleDetectedException;
 
 import com.google.common.collect.Lists;
 
-public class MailboxHierarchySorter {
+public class CollectionHierarchySorter<T, Id> {
```
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/CollectionHierarchySorter.java]: Sorry to be that picky, but the naming does not "click".
May I suggest HierarchicalCollections as a broader helper or SortingHierarchicalCollections

Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/CollectionHierarchySorter.java]: Ok

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/SetMailboxesResponse.java (hunk 8 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/SetMailboxesResponse.java]: OK

**Comment by fvn-linagora** [comment]
:+1 (let's :shipit: !)

**Pull request merged**